File: WEB/src/utils/api/files.js
import axios from "axios";
import config from "../../../public/config.json";
import { getCookie } from "../utils";

const API_BASE = config.APIURL;
const CHUNK_SIZE = 5 * 1024 * 1024;

export const listFiles = async (path) =>
  axios.get(`${API_BASE}/files/list/${encodeURIComponent(path)}`, {
    withCredentials: true,
    headers: {
      "X-CSRF-TOKEN": getCookie("csrf_access_token"),
    },
  });

export const addFavFile = async (path) => {
  axios.post(
    `${API_BASE}/files/add-favourite?path=${path}`,
    {},
    {
      withCredentials: true,
      headers: {
        "X-CSRF-TOKEN": getCookie("csrf_access_token"),
      },
    }
  );
};

export const rmFavFile = async (path) => {
  axios.post(
    `${API_BASE}/files/rm-favourite?path=${path}`,
    {},
    {
      withCredentials: true,
      headers: {
        "X-CSRF-TOKEN": getCookie("csrf_access_token"),
      },
    }
  );
};

export const getFile = async (path) => {
  const response = await axios.get(`${API_BASE}/files/get/${path}`, {
    withCredentials: true,
    headers: {
      "X-CSRF-TOKEN": getCookie("csrf_access_token"),
    },
    responseType: "blob",
  });
  const blob = response.data;
  const url = URL.createObjectURL(blob);
  return url;
};

export const copyFile = async (path, copy_path) => {
  await axios.post(
    `${API_BASE}/files/copy?file_path=${path}&copy_path=${copy_path}`,
    {},
    {
      withCredentials: true,
      headers: {
        "X-CSRF-TOKEN": getCookie("csrf_access_token"),
      },
    }
  );
};

export const getVideoThumbnail = async (path, time = 0.5) => {
  const response = await axios.get(`${API_BASE}/files/thumbnail/${path}`, {
    params: { time },
    withCredentials: true,
    headers: {
      "X-CSRF-TOKEN": getCookie("csrf_access_token"),
    },
    responseType: "blob",
  });

  return URL.createObjectURL(response.data);
};

export const moveFile = async (path, movePath) => {
  const response = await axios.post(
    `${API_BASE}/files/move`,
    { path: path, move_path: movePath },
    {
      withCredentials: true,
      headers: {
        "X-CSRF-TOKEN": getCookie("csrf_access_token"),
      },
    }
  );
};

export const uploadFileApi = async (path, file, onProgress) => {
  const totalChunks = Math.ceil(file.size / CHUNK_SIZE);
  const uploadId = Date.now() + "-" + file.name;

  let uploadedBytes = 0;

  for (let chunkIndex = 0; chunkIndex < totalChunks; chunkIndex++) {
    const start = chunkIndex * CHUNK_SIZE;
    const end = Math.min(start + CHUNK_SIZE, file.size);
    const chunk = file.slice(start, end);

    const formData = new FormData();
    formData.append("file", chunk, file.name);
    formData.append("upload_id", String(uploadId));
    formData.append("chunk_index", String(chunkIndex));
    formData.append("total_chunks", totalChunks);
    formData.append("filename", file.name);
    formData.append("path", path);

    await axios.post(`${API_BASE}/files/upload`, formData, {
      withCredentials: true,
      headers: {
        "X-CSRF-TOKEN": getCookie("csrf_access_token"),
      },
      onUploadProgress: (progressEvent) => {
        if (progressEvent.total && onProgress) {
          const chunkProgress = progressEvent.loaded / progressEvent.total;
          const totalProgress =
            ((uploadedBytes + progressEvent.loaded) / file.size) * 100;
          onProgress(Math.min(100, Math.round(totalProgress)));
        }
      },
    });

    uploadedBytes += chunk.size;
  }
  const formData = new FormData();
  formData.append("upload_id", uploadId);
  formData.append("total_chunks", totalChunks);
  formData.append("filename", file.name);
  formData.append("path", path);

  const response = await axios.post(
    `${API_BASE}/files/complete-upload`,
    formData,
    {
      withCredentials: true,
      headers: {
        "X-CSRF-TOKEN": getCookie("csrf_access_token"),
      },
    }
  );

  if (onProgress) onProgress(100);

  return response.data;
};

export const renameFile = async (path, newName) => {
  return axios.post(
    `${API_BASE}/files/rename`,
    { path: path, new_name: newName },
    {
      withCredentials: true,
      headers: {
        "X-CSRF-TOKEN": getCookie("csrf_access_token"),
      },
    }
  );
};

export const deleteFile = async (path) => {
  try {
    await axios.delete(`${API_BASE}/files/delete`, {
      params: { path },
      withCredentials: true,
      headers: {
        "X-CSRF-TOKEN": getCookie("csrf_access_token"),
      },
    });
  } catch (err) {
    console.error("Ошибка deleteFile:", err);
  }
};

export const downloadFile = async (path) => {
  const a = document.createElement("a");
  a.href = `${API_BASE}/files/download/${path}`;
  a.download = path.split("/").pop();
  a.click();
};

File: API/src/routers/files.py
from fastapi import UploadFile, File, Request, Form
from fastapi import APIRouter, UploadFile, Request, Response
from fastapi.responses import StreamingResponse
from src.schemas.response.files import *
from src.schemas.request.files import *
from src.config import Config
from fastapi import Depends
from src.utils.validators import validate_auth
from src import services

router = APIRouter(
    prefix="/files", 
    tags=["files"], 
    dependencies=[Depends(validate_auth)]
    )
config = Config()

@router.get("/download/{path:path}")
async def download_file(path: str, request: Request) -> DownloadFileErrorResponse:
    return await services.download_file(path, request)

@router.get("/list/{path:path}", response_model=ListFilesResponse)
async def list_files(path: str, request: Request) -> ListFilesResponse:
    return await services.list_files(path, request)


@router.get("/read/{path:path}", response_model=ReadFileResponse)
async def read_file(path: str, request: Request) -> ReadFileResponse:
    return await services.read_file(path, request)


@router.get("/get/{path:path}")
async def get_file(
    path: str,
    request: Request,
    width: int = None,
) -> StreamingResponse:
    return await services.get_file(path, request, width)

@router.get("/thumbnail/{path:path}")
async def thumbnail(
    request: Request,
    path: str,
    time: float = 0.5,
    width: int = 200,
):
    data = GenVideoThumbRequest(path=path, time=time, width=width)
    return await services.gen_video_thumb(data, request)

@router.post("/add-favourite", response_model=AddFavouriteResponse)
async def add_fav(path: str, request: Request) -> AddFavouriteResponse:
    return await services.add_fav_file(path, request)

@router.post("/rm-favourite", response_model=AddFavouriteResponse)
async def remove_fav(path: str, request: Request) -> AddFavouriteResponse:
    return await services.remove_fav_file(path, request)

@router.post("/upload")
async def upload_files(
    request: Request,
    file: UploadFile = File(...),
    upload_id: str = Form(...),
    chunk_index: int = Form(...),
    path: str = Form(...)
):
    return await services.upload_chunk(request, file, upload_id, chunk_index, path)

@router.post("/complete-upload")
async def complete_upload(
    request: Request,
    upload_id: str = Form(...),
    total_chunks: int = Form(...),
    filename: str = Form(...),
    path: str = Form("/"),
):
    return await services.complete_upload(
        request, upload_id, total_chunks, filename, path
    )

@router.post("/rename", response_model=RenameFileResponse)
async def rename_file(data: RenameFileRequest, request: Request) -> RenameFileResponse:
    return await services.rename_file(data, request)

@router.post("/move", response_model=MoveFileResponse)
async def move_file(data: MoveFileRequest, request: Request) -> MoveFileResponse:
    return await services.move_file(data, request)

@router.post("/copy", response_model=CopyFileResponse)
async def copy_file(data: CopyFileRequest, request: Request) -> CopyFileResponse:
    return await services.copy_file(data, request)

@router.delete("/delete", response_model=DeleteFilesResponse)
async def delete_file(path: str, request: Request) -> DeleteFilesResponse:
    return await services.delete_file(path, request)
